presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java

```diff
@@ -0,0 +1,543 @@
+            return immediateFuture(true);
+        }
+
+        List<Page> pages = splitPage(page, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+
+        long rowCount = pages.stream().mapToLong(Page::getPositionCount).sum();
+        checkState(rowCount == page.getPositionCount());
+        totalRowsAdded.addAndGet(rowCount);
+        totalPagesAdded.addAndGet(pages.size());
+
+        return partitions.get(partitionNumber).enqueuePages(pages);
+    }
+
+    @Override
+    public CompletableFuture<BufferResult> get(OutputBufferId outputBufferId, long startingSequenceId, DataSize maxSize)
+    {
+        requireNonNull(outputBufferId, "outputBufferId is null");
+        checkArgument(maxSize.toBytes() > 0, "maxSize must be at least 1 byte");
+
+        return partitions.get(outputBufferId.getId()).getPages(startingSequenceId, maxSize);
+    }
+
+    @Override
+    public void abort(OutputBufferId bufferId)
+    {
+        requireNonNull(bufferId, "bufferId is null");
+
+        partitions.get(bufferId.getId()).destroy();
+
+        checkFlushComplete();
+    }
+
+    @Override
+    public void setNoMorePages()
+    {
+        state.compareAndSet(OPEN, NO_MORE_PAGES);
+        state.compareAndSet(NO_MORE_BUFFERS, FLUSHING);
+        memoryManager.setNoBlockOnFull();
+
+        partitions.forEach(Partition::setNoMorePages);
+
+        checkFlushComplete();
+    }
+
+    @Override
+    public void destroy()
+    {
+        // ignore destroy if the buffer already in a terminal state.
+        if (state.setIf(FINISHED, oldState -> !oldState.isTerminal())) {
+            partitions.forEach(Partition::destroy);
+            memoryManager.setNoBlockOnFull();
+        }
+    }
+
+    @Override
+    public void fail()
+    {
+        // ignore fail if the buffer already in a terminal state.
+        if (state.setIf(FAILED, oldState -> !oldState.isTerminal())) {
+            memoryManager.setNoBlockOnFull();
+        }
+
+        // DO NOT destroy buffers.  The coordinator manages the teardown of failed queries.
+    }
+
+    private void checkFlushComplete()
+    {
+        if (state.get() != FLUSHING) {
+            return;
+        }
+
+        for (Partition partition : partitions) {
+            if (!partition.isFinished()) {
+                return;
+            }
+        }
+        destroy();
+    }
+
+    @ThreadSafe
+    private static final class Partition
+    {
+        private final String taskInstanceId;
+        private final int partitionNumber;
+        private final OutputBufferId bufferId;
+        private final StateMachine<BufferState> state;
+        private final SharedBufferMemoryManager memoryManager;
+
+        private final AtomicLong rowsAdded = new AtomicLong();
+        private final AtomicLong pagesAdded = new AtomicLong();
+
+        private final AtomicLong bufferedBytes = new AtomicLong();
+
+        private final AtomicLong currentSequenceId = new AtomicLong();
+        private final AtomicBoolean finished = new AtomicBoolean();
```

Comment:
The meaning of `finished` is not apparent to me. It appears like all functions except `getInfo` should be no-ops once this flag is set. But then, the behavior of some functions here doesn't match my theory.